ngraph/frontend/generic/src/frontend_manager.cpp

```diff
@@ -0,0 +1,380 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <ngraph/except.hpp>
+
+#include "frontend_manager/frontend_manager.hpp"
+#include "plugin_loader.hpp"
+
+namespace ngraph
+{
+    namespace frontend
+    {
+
+#define FRONT_END_NOT_IMPLEMENTED(NAME) throw std::runtime_error(#NAME " is not implemented for this FrontEnd class")
+#define FRONT_END_ASSERT(EXPRESSION) \
+        { if (!(EXPRESSION)) throw "AssertionFailed"; }
+
+        //----------- FrontEndManager ---------------------------
+        class FrontEndManager::Impl
+        {
+            std::vector<PluginHandle> m_loadedLibs; // must be a first class member (destroyed last)
+            std::map<std::string, FrontEndFactory> m_factories;
+        public:
+            Impl(const std::string& dirName)
+            {
+                registerPlugins(dirName);
+            }
+
+            ~Impl() = default;
+
+            FrontEnd::Ptr loadByFramework(const std::string& framework, FrontEndCapabilities fec)
+            {
+                FRONT_END_ASSERT(m_factories.count(framework))
+                return m_factories[framework](fec);
+            }
+
+            std::vector<std::string> availableFrontEnds() const
+            {
+                std::vector<std::string> keys;
+
+                std::transform(m_factories.begin(), m_factories.end(),
+                               std::back_inserter(keys),
+                               [](const std::pair<std::string, FrontEndFactory>& item)
+                               {
+                                   return item.first;
+                               });
+                return keys;
+            }
+
+            FrontEnd::Ptr loadByModel(const std::string& path, FrontEndCapabilities fec)
+            {
+                FRONT_END_NOT_IMPLEMENTED(loadByModel);
+            }
+
+            void registerFrontEnd(const std::string& name, FrontEndFactory creator)
+            {
+                m_factories.insert({name, creator});
+            }
+
+        private:
+            void registerPlugins(const std::string& dirName)
+            {
+                auto registerFromDir = [&](const std::string& dir)
+                {
+                    auto plugins = loadPlugins(dir);
+                    for (auto& plugin : plugins)
+                    {
+                        registerFrontEnd(plugin.m_pluginInfo.m_name, plugin.m_pluginInfo.m_creator);
+                        m_loadedLibs.push_back(std::move(plugin.m_libHandle));
+                    }
+                };
+                if (!dirName.empty())
+                {
+                    registerFromDir(dirName);
+                    registerFromDir(dirName+"/lib");
+                }
+                registerFromDir("./frontends");
+                registerFromDir("./lib/frontends");
```

Comment:
please, discuss with @asomsiko  whether it's safe to auto preload all libraries from such hardcoded dirs?

Reply:
Changed it to read 'OV_FRONTEND_PATH' environment variable